server/grpc_service.go

```diff
@@ -689,6 +728,83 @@ func (s *heartbeatServer) Recv() (*pdpb.RegionHeartbeatRequest, error) {
 	return req, nil
 }
 
+// ReportBuckets implements gRPC PDServer
+func (s *GrpcServer) ReportBuckets(stream pdpb.PD_ReportBucketsServer) error {
+	var (
+		server            = &bucketHeartbeatServer{stream: stream}
+		forwardStream     pdpb.PD_ReportBucketsClient
+		cancel            context.CancelFunc
+		lastForwardedHost string
+		errCh             chan error
+	)
+	defer func() {
+		if cancel != nil {
+			cancel()
+		}
+	}()
+	for {
+		request, err := server.Recv()
+		if err == io.EOF {
+			return nil
+		}
+		forwardedHost := getForwardedHost(stream.Context())
+		if !s.isLocalRequest(forwardedHost) {
+			if forwardStream == nil || lastForwardedHost != forwardedHost {
+				if cancel != nil {
+					cancel()
+				}
+				client, err := s.getDelegateClient(s.ctx, forwardedHost)
+				if err != nil {
+					return err
+				}
+				log.Info("create bucket report forward stream", zap.String("forwarded-host", forwardedHost))
+				forwardStream, cancel, err = s.createReportBucketsForwardStream(client)
+				if err != nil {
+					return err
+				}
+				lastForwardedHost = forwardedHost
+				errCh = make(chan error, 1)
+				go forwardReportBucketClientToServer(forwardStream, server, errCh)
+			}
+			if err := forwardStream.Send(request); err != nil {
+				return errors.WithStack(err)
+			}
+
+			select {
+			case err := <-errCh:
+				return err
+			default:
+			}
+			continue
+		}
+		rc := s.GetRaftCluster()
+		if rc == nil {
+			resp := &pdpb.ReportBucketsResponse{
+				Header: s.notBootstrappedHeader(),
+			}
+			err := server.Send(resp)
+			return errors.WithStack(err)
+		}
+		if err := s.validateRequest(request.GetHeader()); err != nil {
+			return err
+		}
+		start := time.Now()
+		bucketReportCounter.WithLabelValues("report", "recv").Inc()
+		buckets := request.GetBuckets()
+		if buckets == nil || len(buckets.Keys) == 0 {
+			continue
+		}
+		err = rc.HandleBucketHeartbeat(request.Buckets)
+		if err != nil {
+			log.Warn("update bucket failed", zap.Error(err))
```

Comment:
Prefer to remove this log or change the level.

Reply:
fix